src/ir_passes/mono/monosig.rs

```diff
@@ -0,0 +1,753 @@
+use super::Monomorphize;
+use crate::ir::{self, Ctx, Foreign, MutCtx};
+use itertools::Itertools;
+use std::collections::HashMap;
+
+/// Used for monomorphizing a component's signature when we add it to the queue.
+/// Any functions needed for monomorphizing the signature are located here - the rest are
+/// in MonoDeferred.
+pub struct MonoSig {
+    /// The name of the monomorphized component
+    pub base: ir::Component,
+    /// The underlying component's idx
+    pub underlying_idx: ir::CompIdx,
+    /// Mapping from parameters in the underlying component to their constant bindings.
+    pub binding: ir::Bind<ir::ParamIdx, u64>,
+    /// Queue of facts to monomorphize
+    pub fact_queue: Vec<ir::Fact>,
+
+    /// Mapping of underlying invokes (and how many times we've seen it) to base invokes
+    pub inv_map: HashMap<(ir::InvIdx, u32), ir::InvIdx>,
+    /// Mapping from underlying invokes to how many times we've seen it so far
+    pub inv_counter: HashMap<ir::InvIdx, u32>,
+    /// Mapping of underlying instances (and how many times we've seen it) to base instances
+    pub inst_map: HashMap<(ir::InstIdx, u32), ir::InstIdx>,
+    /// Maapping from underlying instances to how many times we've seen it so far
+    pub inst_counter: HashMap<ir::InstIdx, u32>,
+    /// Hold onto the current PortIdx being handled
+    pub curr_port: Option<ir::PortIdx>,
+    /// Invokes we've already handled, such as when we see an invoke's port before the actual invoke
+    /// so we don't create duplicates
+    pub handled_invokes: Vec<ir::InvIdx>,
+    pub handled_instances: Vec<ir::InstIdx>,
```

Comment:
We discussed making it an IR invariant to have use of ports be dominated by their definition. Can you open an issue to track the creation of a pass that does this?

Reply:
#245